perf-tool/src/exception.cpp

```diff
@@ -0,0 +1,174 @@
+                break;
+            } else {
+                lineNumber = lineTable[i].line_number;
+            }
+        }
+        jdata["callingMethodLineNumber"] = lineNumber;  // record line number of calling method
+    }
+
+    /* Get jclass object of calling method*/
+    err = jvmtiEnv->GetMethodDeclaringClass(method, &klass);
+    check_jvmti_error(jvmtiEnv, err, "Unable to get method declaring class.\n");
+
+    /* Get source file name */
+    err = jvmtiEnv->GetSourceFileName(klass, &fileName);
+    if (err == JVMTI_ERROR_NONE) {
+        jdata["callingMethodSourceFile"] = fileName;
+    }
+
+    // Get information from stack
+    if (backTraceEnabled) { // only run when backtrace is enabled
+        if (exceptionSampleCount % exceptionSampleRate == 0) {
+            int numFrames = 5;
+            jvmtiFrameInfo frames[numFrames];
+            jint count;
+            auto jMethods = json::array();
+
+            err = jvmtiEnv->GetStackTrace(thread, 0, numFrames,
+                                        frames, &count);
+            if (err == JVMTI_ERROR_NONE && count >= 1) {
+                json jMethod;
+                for (int i = 0; i < count; i++) {
+                    /* Get method name */
+                    err = jvmtiEnv->GetMethodName(frames[i].method,
+                                        &methodName, NULL, NULL);
+                    if (err == JVMTI_ERROR_NONE) {
+                        jMethod["methodName"] = methodName;
+                    }
+
+                    err = jvmtiEnv->GetLineNumberTable(frames[i].method, &lineCount, &lineTable);
+                    if (err == JVMTI_ERROR_NONE) { // Find line
+                        lineNumber = lineTable[0].line_number;
+                        for (int j = 1; j < lineCount; j++) {
+                            if (frames[i].location < lineTable[i].start_location) {
+                                break;
+                            } else {
+                                lineNumber = lineTable[i].line_number;
+                            }
+                        }
+
+                        jMethod["lineNumber"] = lineNumber;
+                    }
+
+                    /* Get jclass object of calling method*/
+                    err = jvmtiEnv->GetMethodDeclaringClass(frames[i].method, &klass);
+                    check_jvmti_error(jvmtiEnv, err, "Unable to get method declaring class.\n");
+
+                    /* Get file name */
+                    err = jvmtiEnv->GetSourceFileName(klass, &fileName);
+                    if (err == JVMTI_ERROR_NONE) {
+                        jMethod["fileName"] = fileName;
+                    }
+
+                    jMethods.push_back(jMethod);
+                }
+            }
+
+            jdata["backtrace"] = jMethods;
+        }
+
+        exceptionSampleCount++;
+    }
+
+    jdata["numExceptions"] = numExceptions;
+
+    err = jvmtiEnv->Deallocate((unsigned char*)fileName);
```

Comment:
I am note versed in jvmti, but this code makes me thing that the jvmti code allocates some strings every time you do `jvmtiEnv->GetMethodName` and things of this sort. If that's the case we need to do a deallocate after every such operation, not just here at the end.